mapper/pchain/tx_parser.go

```diff
@@ -0,0 +1,740 @@
+	ops := newTxOps(t.isConstruction)
+
+	err := t.insToOperations(ops, txType, tx.Ins, OpTypeInput)
+	if err != nil {
+		return nil, err
+	}
+
+	err = t.outsToOperations(ops, txType, txID, tx.Outs, OpTypeOutput, mapper.PChainNetworkIdentifier)
+	if err != nil {
+		return nil, err
+	}
+
+	return ops, nil
+}
+
+func (t *TxParser) insToOperations(
+	inOps *txOps,
+	opType string,
+	txIns []*avax.TransferableInput,
+	metaType string,
+) error {
+	status := types.String(mapper.StatusSuccess)
+	if t.isConstruction {
+		status = nil
+	}
+
+	for _, in := range txIns {
+		metadata := &OperationMetadata{
+			Type: metaType,
+		}
+
+		if transferInput, ok := in.In.(*secp256k1fx.TransferInput); ok {
+			metadata.SigIndices = transferInput.SigIndices
+		}
+
+		opMetadata, err := mapper.MarshalJSONMap(metadata)
+		if err != nil {
+			return err
+		}
+
+		// If dependency txs are provided, which is the case for /block endpoints
+		// check whether the input UTXO is multisig. If so, skip it.
+		if t.dependencyTxs != nil {
+			isMultisig, err := t.isMultisig(in.UTXOID)
+			if err != nil {
+				return errFailedToCheckMultisig
+			}
+			if isMultisig {
+				continue
+			}
+		}
+
+		utxoID := in.UTXOID.String()
+		account, ok := t.inputTxAccounts[utxoID]
+		if !ok {
+			return errNoMatchingInputAddresses
+		}
+
+		inputAmount := new(big.Int).SetUint64(in.In.Amount())
+		inOp := &types.Operation{
+			OperationIdentifier: &types.OperationIdentifier{
+				Index: int64(inOps.Len()),
+			},
+			Type:    opType,
+			Status:  status,
+			Account: account,
+			// Negating input amount
+			Amount: mapper.AtomicAvaxAmount(new(big.Int).Neg(inputAmount)),
+			CoinChange: &types.CoinChange{
+				CoinIdentifier: &types.CoinIdentifier{
+					Identifier: utxoID,
+				},
+				CoinAction: types.CoinSpent,
+			},
+			Metadata: opMetadata,
+		}
+
+		inOps.Append(inOp, metaType)
+	}
+	return nil
+}
+
+func (t *TxParser) outsToOperations(
+	outOps *txOps,
+	opType string,
+	txID ids.ID,
+	txOut []*avax.TransferableOutput,
+	metaType string,
+	chainIDAlias string,
+) error {
+	outIndexOffset := outOps.OutputLen()
+	status := types.String(mapper.StatusSuccess)
+	if t.isConstruction {
+		status = nil
+	}
+
+	for outIndex, out := range txOut {
+		transferOut := out.Out
+
+		if lockOut, ok := transferOut.(*stakeable.LockOut); ok {
+			transferOut = lockOut.TransferableOut
+		}
+
+		transferOutput, ok := transferOut.(*secp256k1fx.TransferOutput)
+		if !ok {
+			return errOutputTypeAssertion
+		}
+
+		// Rosetta cannot handle multisig at the moment. In order to pass data validation,
+		// we treat multisig outputs like a burn and inputs line a mint and therefore
+		// not include them in the operations
+		//
+		// Additionally, it is possible to have outputs without any addresses
+		// (e.g. https://testnet.avascan.info/blockchain/p/block/81016)
+		//
+		// therefore we skip parsing operations unless there is exactly 1 address
```

Comment:
I don't really know what the implications of this are but...is this OK? Is Rosetta going to report incorrect balances?

Reply:
It is correct if a particular account uses multisig. Rosetta operations are currently tied to 1 account. We also couldn't figure out how to model which account's balance we are counting a given UTXO against. 

With that in mind, this is the best trade-off we could think of short of making a Rosetta spec change.